### PR TITLE
PackageResolver Refactor

### PR DESCRIPTION
**Pull request by cloudRoutine**
Started working on this in an attempt to address https://github.com/fsprojects/Paket/issues/2030

I wasn't able to get the step function to the point where its tail recursive, but its much closer than it was before and it no longer relies on mutation across iterations.

I think it should be possible by adding a third case to 
```fsharp
type private Stage =
    | Outer of conflictState : ConflictState
    | Inner of conflictState : ConflictState
```
and using it as an argument for `step` to direct it down one of three pathways for the start of a step, an outer loop, or an inner loop, although when I tried coding that up I couldn't get it to work.

I think this is the block of code that was causing the issues when I took that approach
```fsharp
                    match conflictState.Status with
                    | Resolution.Conflict(_,conflicts,_,_)
                        when
                            (Set.isEmpty conflicts |> not)
                            && nextStep.CurrentResolution.Count > 1
                            && not (conflicts |> Set.exists (fun r ->
                                r = currentRequirement
                                || r.Graph |> List.contains currentRequirement)) ->
                        let flags = StepFlags(flags.Ready,flags.UseUnlisted,flags.HasUnlisted,true,flags.FirstTrial)
                        stepLoop flags (Inner (conflictState))
                    | _ ->
                        stepLoop flags (Inner (conflictState))
```

but I'm not totally sure.

### CODE REVIEW

File: src/Paket.Core/PackageResolver.fs [hunk 1]
```diff
@@ -94,90 +99,111 @@ type ResolverStep = {
...
+    let getConflicts (res:Resolution) =
+        match res with
+        | Resolution.Ok _ -> []
+        | Resolution.Conflict (currentStep,_,lastPackageRequirement,_) ->
             currentStep.ClosedRequirements
             |> Set.union currentStep.OpenRequirements
             |> Set.add lastPackageRequirement
-            |> Seq.filter (fun x -> x.Name = lastPackageRequirement.Name)
-            |> Seq.sortBy (fun x -> x.Parent)
+            |> Seq.filter ^ fun x -> x.Name = lastPackageRequirement.Name
```
Comment (forki) [inline src/Paket.Core/PackageResolver.fs]: please don't

**Comment by forki** [comment]
this is really really hard to review- can you please split it in two PRs. one with the resolver change and one with all the cosmetical changes? thanks

**Comment by cloudRoutine** [comment]
@forki this fine?

**Comment by forki** [comment]
it seems the new resolver is not resolving correctly tests like ##1254 are red. can you confirm?

**Comment by cloudRoutine** [comment]
a bug got reintroduced when I redid the commits

**Comment by forki** [comment]
#1883 install FSharp.Core from Chessie fails!?

**Comment by cloudRoutine** [comment]
I've been having issues running the integration tests on my machine

**Comment by forki** [comment]
yeah these new packages are fucking up everything. MS is clearly out of control here.

**Comment by cloudRoutine** [comment]
even when I try to run the [integration tests on the core3 branch they don't complete](https://gist.github.com/cloudRoutine/347eb5b83b6078158f825183ad3b22e9)

**Comment by forki** [comment]
do you have VS open at that time? or something else that keeps an handle on the nupkg?

**Comment by cloudRoutine** [comment]
I had all of that stuff closed at the time. The first time I got those errors I thought that might be the source but when I tried again with everything off it still happened.

I get these errors when I run the tests just in VS too

**Comment by forki** [comment]
And from fake build?

Am 23.11.2016 16:50 schrieb "Jared Hester" <notifications@github.com>:

**Comment by cloudRoutine** [comment]
the log in the gist is from the FAKE build

**Comment by forki** [comment]
Ok can you please try to find out which tool is locking it? Sorry for all the work, but this whole .net standard conversion is a real mess.

**Comment by forki** [comment]
https://github.com/fsprojects/Paket/pull/1977 seems to work on appveyor. (travis is expected to fail)

**Comment by cloudRoutine** [comment]
The readme might need a note to install F# 3.0, without it `#144` will fail with



But even after I installed F# 3.0

**Comment by cloudRoutine** [comment]
now I'm really confused about where this error is coming from
when I run the contents of test 1883 in FSI

**Comment by cloudRoutine** [comment]
A more detailed output

File: src/Paket.Core/PackageResolver.fs [hunk 2]
```diff
@@ -257,313 +283,489 @@ type UpdateMode =
...
+
+                match getExploredPackage packageDetails getPackageDetailsF exploredPackages with
+                | None ->
+                    stepLoop flags (Inner(conflictState))
+                | Some exploredPackage ->
+                    let hasUnlisted = exploredPackage.Unlisted || flags.HasUnlisted
+
+                    let flags = StepFlags(flags.Ready,flags.UseUnlisted,hasUnlisted,flags.ForceBreak,flags.FirstTrial)
+                    if exploredPackage.Unlisted && not flags.UseUnlisted then
+                        tracefn "     unlisted"
```
Comment (cloudRoutine) [inline src/Paket.Core/PackageResolver.fs]: now i'm pretty sure the bug is here, where I last made a change :facepalm:
which I'd had as 
```fsharp
                    if exploredPackage.Unlisted && not flags.UseUnlisted then
                        tracefn "     unlisted"
                        stepLoop flags (Inner(conflictState))
                    else
```
before, but that caused an infinite loop printing `unlisted`

@forki it'd be really useful for whoever has to work on this next, if you or someone else with a deeper understanding of what's going on in this module could add some comments explaining the processes. I'd do it myself, but I don't think I have a firm enough grasp and I wouldn't want to put down something that's wrong

**Comment by cloudRoutine** [comment]
@forki finally passed on appveyor 🎉

**Comment by cloudRoutine** [comment]
anything else I need to do for this?

**Comment by forki** [comment]
thanks so much. let's what the real world thinks about it ;-)

**Pull request merged**